### PR TITLE
CI: Implement "make check".

### PR DESCRIPTION
**Pull request by infrastation**
This revision works on my Linux as expected, let's see if this holds for the other OSes.

### CODE REVIEW

**Comment by infrastation** [comment]
This needs a bit more work. One unexpected result is that DECnet bytecode has some immediate constants byte-swapped on MIPS64.

**Comment by infrastation** [comment]
On a separate note, it turns out musl libc hard-codes a list of protocols known by `getprotobyname()` and the list does not include "sctp", so the presence of "sctp" in `/etc/protocols` on Alpine Linux has no effect.

**Comment by guyharris** [comment]
> This needs a bit more work. One unexpected result is that DECnet bytecode has some immediate constants byte-swapped on MIPS64.

And, I suspect, on z/Architecture as well, if tests were being done there.

I.e., that's *probably* a big-endian MIPS machine (I don't know whether anybody other than DEC ran MIPS processors big-endian).

**Comment by guyharris** [comment]
> On a separate note, it turns out musl libc hard-codes a list of protocols known by `getprotobyname()` and the list does not include "sctp",

Thanks, musl developers!

Perhaps if `getprotobyname()` fails on "sctp", maybe we should just override it with our own hard-coded version.

Or maybe we should have a hard-coded table of the non-obscure protocols, check there first and, if that fails, use `getprotobyname()`.

**Comment by infrastation** [comment]
The s390x and MIPS64 Buildbot hosts are big-endian. I vaguely remember some similar endianness-related bytecode discrepancies a while ago.

Here is the difference between `getent protocols` on a Debian 12 box and [musl source code](https://git.musl-libc.org/cgit/musl/tree/src/network/proto.c):
```diff
--- tmp/protocols-glibc.txt	2024-12-08 23:16:40.400502024 +0000
+++ tmp/protocols-musl.txt	2024-12-08 23:17:07.836254053 +0000
@@ -1,26 +1,16 @@
 ah
-ax.25
-dccp
 ddp
 egp
-eigrp
 encap
 esp
-etherip
-ethernet
-fc
 ggp
 gre
-hip
 hmp
-hopopt
 icmp
 idpr-cmtp
 idrp
 igmp
-igp
 ip
-ipcomp
 ipencap
 ipip
 ipv6
@@ -29,29 +19,18 @@
 ipv6-nonxt
 ipv6-opts
 ipv6-route
-isis
 iso-tp4
-l2tp
-manet
-mobility-header
-mpls-in-ip
-mptcp
 ospf
 pim
 pup
+raw
 rdp
-rohc
 rspf
 rsvp
-sctp
-shim6
 skip
 st
 tcp
 udp
-udplite
 vmtp
-vrrp
-wesp
 xns-idp
 xtp
```
Based on that, having a last-resort protocol lookup table may be a good idea.

That said, it may be wrong in the first place to use in a test case expression too many protocols that depend on external resolution. Perhaps just TCP, UDP and ICMP would suffice to prove that the code path to `getprotobyname()` works.

**Comment by infrastation** [comment]
[**pcap-filter**](https://www.tcpdump.org/manpages/pcap-filter.7.html)(7) now documents musl libc peculiarity.

**Comment by infrastation** [comment]
Fixed an OpenBSD-specific bug and made the new shell script code compatible with FreeBSD and illumos shells. Is the `timeout` command available on macOS somehow?

**Comment by guyharris** [comment]
> Is the `timeout` command available on macOS somehow?

macOS doesn't provide a `timeout` command.

With a tiny tweak (replace `sys_nsig` with `NSIG`), the NetBSD timeout.c compiles and appears to work.

**Comment by infrastation** [comment]
A search on the Interwebs indicates GNU coreutils are available as a brew package, but then either it needs to be called as `gtimeout` or `PATH` needs to be customised. Maybe I will just disable the timeout on macOS for now, get the first revision off the ground and then see what to improve next.

**Comment by guyharris** [comment]
Is the goal to limit *real* time taken by a test (e.g., to keep it from hanging forever on a host name lookup) or to limit the *CPU* time taken by a test (e.g., to keep it from getting stuck in an infinite loop)?

If it's the latter, and if those commands are being run by a shell, you could do

    (ulimit -t <CPU time seconds>; <command>)

to set a CPU time limit for the shell in which the command is run.

If it's the former, there's no `setrlimit()` limit for wall-clock time (at least not on macOS, according to the man page), so `ulimit` won't help.

**Comment by infrastation** [comment]
These tests are meant not to depend on any external effects (including DNS), so normally the CPU time should be very close to the wall clock time.

The goal is as follows. Both `timeout` and `ulimit` should be able to detect a true infinite loop using, let's say, a 60 second timeout. The current revision tries to get a bit further and to detect a performance regression — if a one-line expression starts to take 100 times the usual time to compile, it would be useful to detect that before more complex filters become subject to the regression.  This seems to be a matter of getting the timeout right.

One difficulty is that CPU performance varies significantly, one order of magnitude or perhaps more.  Another is that Valgrind usage impacts performance, the worst case seems to be 7.5 seconds per `filtertest` invocation.  And yet another is that it is not uncommon to have other workload running in parallel, so the test, generally speaking may be not guaranteed an idle CPU core. In the latter case `ulimit` would be more appropriate (if portability does not get in the way).

So ideally perhaps using both (`timeout` with a large value and `ulimit` with a smaller value) would make the most sense.

**Comment by infrastation** [comment]
Whilst trying to write up a comment to explain why `timeout` is inconvenient to get on macOS, I ran out of patience and made it possible to use `gtimeout` instead. One consequence of this is that Autoconf has automatically switched to `ginstall`, so the build is now less representative of macOS.

**Comment by infrastation** [comment]
`INSTALL=/usr/bin/install` should force the native version, let me try that.

**Comment by infrastation** [comment]
Yes, that works as expected:
```
checking for a BSD-compatible install... /usr/bin/install
```

File: .cirrus.yml [hunk 1]
```diff
@@ -46,9 +46,11 @@ macos_task:
     image: ghcr.io/cirruslabs/macos-runner:sonoma # last 3 versions of Xcode
   env:
     MAKEFLAGS: '-j 4' # macOS VMs run on 4 cores
+    TIMEOUT_bin: gtimeout
```
[the remark this answers is not in the document]
Comment (infrastation) [reply .cirrus.yml:49]: Yes, thank you. The only way I can explain why the typo has not broken the build is that installing coreutils makes `timeout` available in the path. In this case the variable is not required. Let's try that too...

[the remark this answers is not in the document]
Comment (infrastation) [reply .cirrus.yml]: Yes, it works without the variable.

**Pull request merged**